tests/unit/s2n_self_talk_io_mem_test.c

```diff
@@ -0,0 +1,216 @@
+        EXPECT_EQUAL(client_conn->max_outgoing_fragment_length, S2N_LARGE_FRAGMENT_LENGTH);
+        EXPECT_EQUAL(server_conn->max_outgoing_fragment_length, S2N_SMALL_FRAGMENT_LENGTH);
+        EXPECT_EQUAL(client_conn->out.blob.size, S2N_TLS13_MAX_RECORD_LEN_FOR(S2N_LARGE_FRAGMENT_LENGTH));
+        EXPECT_EQUAL(server_conn->out.blob.size, S2N_TLS13_MAX_RECORD_LEN_FOR(S2N_LARGE_FRAGMENT_LENGTH));
+
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+    /* Allocate output buffer to value negotiated with max_fragment_length extension */
+    {
+        const s2n_max_frag_len mfl_code = S2N_TLS_MAX_FRAG_LEN_2048;
+        const uint16_t expected_mfl = 2048;
+
+        struct s2n_config *config_for_mfl = s2n_config_new();
+        EXPECT_NOT_NULL(config_for_mfl);
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config_for_mfl, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config_for_mfl));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config_for_mfl, chain_and_key));
+        EXPECT_SUCCESS(s2n_config_send_max_fragment_length(config_for_mfl, mfl_code));
+        EXPECT_SUCCESS(s2n_config_accept_max_fragment_length(config_for_mfl));
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config_for_mfl));
+
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config_for_mfl));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair;
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Set connections to use smaller fragment size */
+        EXPECT_SUCCESS(s2n_connection_prefer_low_latency(client_conn));
+        EXPECT_SUCCESS(s2n_connection_prefer_throughput(server_conn));
+
+        /* Do handshake */
+        EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server_conn, client_conn, SERVER_CERT));
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+
+        /* Output memory allocated according to max fragment length. */
+        EXPECT_EQUAL(client_conn->max_outgoing_fragment_length, S2N_SMALL_FRAGMENT_LENGTH);
+        EXPECT_EQUAL(server_conn->max_outgoing_fragment_length, expected_mfl);
+        EXPECT_EQUAL(client_conn->out.blob.size, S2N_TLS_MAX_RECORD_LEN_FOR(S2N_SMALL_FRAGMENT_LENGTH));
+        EXPECT_EQUAL(server_conn->out.blob.size, S2N_TLS13_MAX_RECORD_LEN_FOR(expected_mfl));
```

Comment:
Just for my understanding: I would have expected server_conn->out to be equal to the default still since expected_mfl=2048 and `s2n_connection_set_max_fragment_length` reallocs when conn->out is smaller (but I didn't think it actually shrunk things). Is there another place where this is shrank from the default? This is related to the comment I wrote on `s2n_connection_set_max_fragment_length` so I think I am missing something about how we initialize conn->out and where it shrinks/grows.

Reply:
from offline convo: this is intentional and works because the first call to writev() happens after MFL negotiation happens.